### PR TITLE
[Model][Decoder]:Add decoder structure.

### PR DESCRIPTION
**Pull request by LingerDavid**
Add decoder structure based on the transformer.

### CODE REVIEW

File: nn_modules/decoder/.gitignore
Comment (eleven-yun) [inline nn_modules/decoder/.gitignore:1]: Shall we move this file to the root?

File: nn_modules/decoder/decoder.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
+'''
+Author: diwei diwei@qcraft.ai
```
Comment (eleven-yun) [inline nn_modules/decoder/decoder.py]: I suggest we avoid disclosing one's company information here as this is intended as a public project.

File: nn_modules/decoder/decoder.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+from torch import nn
+from torch import Tensor
+from typing import Optional
+
+
+class TopDecoder(nn.Module):
+    def __init__(self, d_model: int, nhead: int) -> None:
+        super().__init__()
+        self.decoder_layer = nn.TransformerDecoderLayer(
+            d_model, nhead, batch_first=True)
```
Comment (eleven-yun) [inline nn_modules/decoder/decoder.py]: We probably need to align with this option in encoder and decoder.
Is there any advantage that we set batch first?

File: nn_modules/decoder/decoder.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+    def forward(self, x: Tensor, memory: Tensor, mask: Optional[Tensor] = None) -> Tensor:
```
Comment (eleven-yun) [inline nn_modules/decoder/decoder.py]: Not sure if we really need the mask in our specific problem.
My opinion is that if it is not used, then we could simply omit it.

File: nn_modules/decoder/decoder_test.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+FilePath: /top_eleven/nn_modules/decoder/decoder_test.py
+Description: 
+
+Copyright (c) 2024 by ${git_name_email}, All Rights Reserved. 
+'''
+
+
+import unittest
+import torch
+from decoder import *
```
Comment (eleven-yun) [inline nn_modules/decoder/decoder_test.py]: I have the impression that import * is not a recommended practice.

**Review by eleven-yun — CHANGES_REQUESTED** [review]
Looks good except for some minor issues.

File: nn_modules/decoder/decoder.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+Copyright (c) 2024 by ${git_name_email}, All Rights Reserved. 
+'''
+from torch import nn
+from torch import Tensor
+from typing import Optional
+
+
+class TopDecoder(nn.Module):
+    def __init__(self, d_model: int, nhead: int) -> None:
```
Comment (BackSpaceShift) [inline nn_modules/decoder/decoder.py]: Do we need multiple decoder layers here?

Comment (LingerDavid) [reply nn_modules/decoder/decoder.py]: It may be necessary, packaging is for future extension and custom operations.

File: nn_modules/decoder/decoder.py (hunk 2 above)
Comment (LingerDavid) [reply nn_modules/decoder/decoder.py]: Modified

File: nn_modules/decoder/decoder_test.py (hunk 4 above)
Comment (LingerDavid) [reply nn_modules/decoder/decoder_test.py]: Make sense.Done.

File: nn_modules/decoder/decoder.py (hunk 1 above)
Comment (LingerDavid) [reply nn_modules/decoder/decoder.py]: Okay.

File: nn_modules/decoder/decoder.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+
+class TopDecoder(nn.Module):
+    def __init__(self, num_layers: int, d_model: int, nhead: int) -> None:
+        super(TopDecoder, self).__init__()
+        self.layers = nn.ModuleList(
+            [nn.TransformerDecoderLayer(d_model, nhead) for _ in range(num_layers)])
+
+    def forward(self, x: Tensor, memory: Tensor) -> Tensor:
+        """Pass the inputs (and mask) through the decoder layer.
```
Comment (eleven-yun) [inline nn_modules/decoder/decoder.py]: we don't have mask anymore so remove it from the docstring, too.

File: nn_modules/decoder/decoder.py [hunk 7]
```diff
@@ -0,0 +1,31 @@
+from torch import nn
+from torch import Tensor
+from typing import Optional
```
Comment (eleven-yun) [inline nn_modules/decoder/decoder.py]: Do we need this import?

**Review by eleven-yun — CHANGES_REQUESTED** [review]
Almost there.

File: nn_modules/decoder/decoder.py (hunk 6 above)
Comment (LingerDavid) [reply nn_modules/decoder/decoder.py]: Done

**Review by eleven-yun — APPROVED** [review]
Nice!

File: nn_modules/decoder/decoder.py (hunk 7 above)
Comment (LingerDavid) [reply nn_modules/decoder/decoder.py]: Remove it already.
